framework for custom content into Lunr indices (#1068)
just-the-docs/just-the-docs: docs/search.md

This command creates the `assets/js/zzzz-search-data.json` file that Jekyll uses to create your search index.
Alternatively, you can create the file manually with [this content]({{ site.github.repository_url }}/blob/main/assets/js/zzzz-search-data.json).

## Custom content for search index

The standard text that is indexed is the page (or post) `.content`, `.title`, and sometimes headers which are already in the `.content`. Other text (Front Matter, data files, etc.) is not indexed. When you want additional text to be indexed, you can customize Just the Docs.

{: .warning }
> Customizing search indices is an advanced feature that requires Javascript and Liquid knowledge.

1. When your site used a previous version of Just the Docs, you must update the file

>    `assets/js/zzzz-search-data.json` using the [above methods](#generate-search-index-when-used-as-a-gem).

Reviewer: @mattxwang I'm a bit puzzled by this bit of the docs update. Surely the _sources_ of a site using JtD as a (remote) theme should _**not**_ have a copy of the file `assets/js/zzzz-search-data.json` unless it is already customising search. Please clarify!

>    `assets/js/zzzz-search-data.json` using the [above methods](#generate-search-index-when-used-as-a-gem).

Author: Good question and I have no answer to that 🤷‍♀️...I have no experience with remote themes. Since it is...remote...I would assume there are no "local" core theme files. Unless they are overriden...if remote themes support that.

>    `assets/js/zzzz-search-data.json` using the [above methods](#generate-search-index-when-used-as-a-gem).

Reviewer: The tests site uses `remote_theme: just-the-docs/just-the-docs`, and I think it's equivalent to `theme: just-the-docs` with a Gemfile that pulls the gem from GitHub.

I find it quite convenient, because I can simply append `@REF` where `REF` can be any tag (release, pre-release, branch, commit) and restart Jekyll, without needing to rerun `bundle`.

And yes, the tests site generally relies on the (remote) theme providing everything in `_includes`, `_layouts`, and `_sass`, but it can override them when needed.

>    `assets/js/zzzz-search-data.json` using the [above methods](#generate-search-index-when-used-as-a-gem).

Reviewer 2: Generally, users of our theme shouldn't have a copy of this; however, in my experience I have seen many, many sites that have copied the file directly into their repo. I had also mostly made this edit originally based off of @diablodale's writing.

(there's some reason that this rake task exists - my understanding is that sometimes, the JSON file doesn't exist. Perhaps I need to do more digging here)

>    `assets/js/zzzz-search-data.json` using the [above methods](#generate-search-index-when-used-as-a-gem).

Author: I found in existing JTD docs https://just-the-docs.github.io/just-the-docs/docs/search/#generate-search-index-when-used-as-a-gem
> If you use Just the Docs as a remote theme, you do not need the following steps. If you use the theme as a gem....do rake

When that is not true, that is a separate issue.
I assume for now it is true.

For gem use, rake needs to be run because those "true" instructions tell me to do that. Any `zzzz-search-data.json` previously created by rake is out-of-date and will not function with this PR -- silently do nothing. This PR needs an updated version of `zzzz-search-data.json` which rake makes.

An experiment: When I `bundle install` and my Gemfile points to `gem "just-the-docs", github: "diablodale/just-the-docs", branch: "dp-release"` it creates a local `/usr/local/bundle/bundler/gems/just-the-docs-5d413480c165/assets/js/zzzz-search-data.json`. I ?assume? that jekyll will use that local system `zzzz` file. Therefore, if a user updated the gem containing this PR, then it should have a new local `zzzz` file. I am not a Ruby or Gem expert...so these are more logical guesses.

>    `assets/js/zzzz-search-data.json` using the [above methods](#generate-search-index-when-used-as-a-gem).

Reviewer 2: Right - my comment is more that, `bundle exec just-the-docs rake search:init` shouldn't be necessary, since we bundle the assets folder as part of the gem. Definitely a separate issue, just wanted to provide some context for the above conversation.

2. Add a new file named `_includes/lunr/custom-data.json` to your site. Insert your custom Liquid
   code that reads the page object at `include.page` then generates custom Javascript
   fields that hold the custom text you want to index. You can verify the fields you output at

>    the generated `assets/js/search-data.json`.

Reviewer: @mattxwang A very minor meta-nit: "ex" could be read as "excluding", and I don't think we should use it as a short form of "e.g." or "for example".

>    the generated `assets/js/search-data.json`.

Author: I'm not confident the recent edits to steps 2 and 3 are clear. As now written, it suggests they are disconnected. "Use 2 for liquid". Use "3 for javascript".
But the truth is the coder must use both. That's how I wrote them tightly together, in sequence, and part of the same numbered list.

This is already a very advanced feature. Anyone using it will know what they are doing. I suspect the majority of these advanced users want to liquid->lunr. And the docs should explain that majority scenario.

If you want to document the ability of two different scenarios:
* liquid->javascript->lunr
* javascript->lunr

...then we need to separate sections and two separate numbered lists.

>    the generated `assets/js/search-data.json`.

Author: > @mattxwang A very minor meta-nit: "ex" could be read as "excluding", and I don't think we should use it as a short form of "e.g." or "for example".
> 
> This seems even more likely to be confusing (at least for European readers) when the colon is omitted, e.g., in "ex front matter, …".

Agreed. "ex" should not be used.  "e.g." or "for example" is the clear and correct usage.

> 3. Add a new file named `_includes/lunr/custom-index.js` to your site. Insert your custom Javascript
>    code that reads your custom Javascript fields and inserts them into the search index.

Reviewer 2: ```suggestion
3. To add JavaScript (or external) data: create a new include at the path `_includes/lunr/custom-index.js` to your site. Add valid JavaScript that creates relevant fields to add to the index. You may want to inspect `assets/js/just-the-docs.js` to better understand the code structure.
```

#### Example

> `_includes/lunr/custom-data.json` custom code reads the page's custom Front Matter `usage` and `examples`
> fields, normalizes the text, and writes the text to custom Javascript `myusage` and `myexamples` fields.

Reviewer 2: ```suggestion
`_includes/lunr/custom-data.json`: this example adds each page's `usage` and `examples` front matter fields, normalizes the text, and writes the text to custom Javascript `myusage` and `myexamples` fields.
```
